packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -278,6 +341,151 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
+    if (composingRegionPrioritized && value.isComposingRangeValid && withComposing) {
+      _addToMappingWithoutOverlap((String value, TextRange range) {
+          final TextStyle composingStyle = style!.merge(
+            const TextStyle(decoration: TextDecoration.underline),
+          );
+          return TextSpan(
+            style: composingStyle,
+            text: value,
+          );
+        },
+        value.composing,
+        rangeSpanMapping,
+        value.text
+      );
+    }
+    // Iterate through TextEditingInlineSpanReplacements, adding non overlapping
+    // to the mapping pointing towards the generated InlineSpan.
+    for (final TextEditingInlineSpanReplacement replacement in replacements) {
+      for (final Match match in replacement.pattern.allMatches(value.text)) {
+        _addToMappingWithoutOverlap(replacement.generator, TextRange(start: match.start, end: match.end), rangeSpanMapping, value.text);
+      }
+    }
+    // If the composing range is out of range for the current text, ignore it to
+    // preserve the tree integrity, otherwise in release mode a RangeError will
+    // be thrown and this EditableText will be built with a broken subtree.
+    //
+    // Add composing region as a replacement to a TextSpan with underline.
+    if (!composingRegionPrioritized && value.isComposingRangeValid && withComposing) {
+      _addToMappingWithoutOverlap((String value, TextRange range) {
+          final TextStyle composingStyle = style!.merge(
+            const TextStyle(decoration: TextDecoration.underline),
+          );
+          return TextSpan(
+            style: composingStyle,
+            text: value,
+          );
+        },
+        value.composing,
+        rangeSpanMapping,
+        value.text
+      );
+    }
+    // Sort the matches by start index. Since no overlapping exists, this is safe.
+    final List<TextRange> sortedRanges = rangeSpanMapping.keys.toList();
+    sortedRanges.sort((TextRange a, TextRange b) => a.start.compareTo(b.start));
+    // Create TextSpans for non-replaced text ranges and insert the replacements spans
+    // for any ranges that are marked to be replaced.
+    final List<InlineSpan> spans = <InlineSpan>[];
+    int previousEndIndex = 0;
+    for (final TextRange range in sortedRanges) {
+      if (range.start > previousEndIndex) {
+        spans.add(TextSpan(text: value.text.substring(previousEndIndex, range.start)));
+      }
+      spans.add(rangeSpanMapping[range]!);
+      previousEndIndex = range.end;
+    }
+    // Add any trailing text as a regular TextSpan.
+    if (previousEndIndex < value.text.length) {
+      spans.add(TextSpan(text: value.text.substring(previousEndIndex, value.text.length)));
+    }
+    return TextSpan(
+      style: style,
+      children: spans,
+    );
+  }
+}
+
+void _addToMappingWithoutOverlap(InlineSpanGenerator generator, TextRange matchedRange, Map<TextRange, InlineSpan> rangeSpanMapping, String text) {
```

Comment:
Can this just be a private (static) method on ReplacementTextEditingController? Why top-level?